flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskOperatorTimerTest.java

```diff
@@ -0,0 +1,144 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Test to verify that timer triggers are run according to operator precedence (combined with yield() at operator level).
+ */
+public class StreamTaskOperatorTimerTest extends TestLogger {
+	private static List<String> events;
+
+	@Test
+	public void testOperatorYieldExecutesSelectedTimers() throws Exception {
+		events = new ArrayList<>();
+		final OneInputStreamTaskTestHarness<Integer, Integer> testHarness = new OneInputStreamTaskTestHarness<>(
+				OneInputStreamTask::new,
+				BasicTypeInfo.INT_TYPE_INFO,
+				BasicTypeInfo.INT_TYPE_INFO);
+
+		testHarness.setupOperatorChain(new OperatorID(), new TestOperatorFactory<>())
+				.chain(new OperatorID(), new TestOperatorFactory<>(), IntSerializer.INSTANCE)
+				.finish();
+
+		testHarness.invoke();
+		testHarness.waitForTaskRunning();
+
+		testHarness.processElement(new StreamRecord<>(42));
+
+		testHarness.endInput();
+		testHarness.waitForTaskCompletion();
+
+		assertThat(events, is(Arrays.asList("Timer:1:1", "Timer:0:1")));
+	}
+
+	private static class TestOperatorFactory<T> implements OneInputStreamOperatorFactory<T, T>, YieldingOperatorFactory<T> {
+		private MailboxExecutor mailboxExecutor;
+
+		@Override
+		public void setMailboxExecutor(MailboxExecutor mailboxExecutor) {
+			this.mailboxExecutor = mailboxExecutor;
+		}
+
+		@Override
+		public <Operator extends StreamOperator<T>> Operator createStreamOperator(
+				StreamTask<?, ?> containingTask,
+				StreamConfig config,
+				Output<StreamRecord<T>> output) {
+			TestOperator<T> operator = new TestOperator<>(config.getChainIndex(), mailboxExecutor);
+			operator.setup(containingTask, config, output);
+			return (Operator) operator;
+		}
+
+		@Override
+		public void setChainingStrategy(ChainingStrategy strategy) {
+		}
+
+		@Override
+		public ChainingStrategy getChainingStrategy() {
+			return ChainingStrategy.ALWAYS;
+		}
+
+		@Override
+		public Class<? extends StreamOperator> getStreamOperatorClass(ClassLoader classLoader) {
+			return TestOperator.class;
+		}
+	}
+
+	private static class TestOperator<T>
+			extends AbstractStreamOperator<T>
+			implements OneInputStreamOperator<T, T> {
+
+		private final transient MailboxExecutor mailboxExecutor;
+		private final int chainIndex;
+		private transient int count;
+
+		TestOperator(int chainIndex, MailboxExecutor mailboxExecutor) {
+			this.chainIndex = chainIndex;
+			this.mailboxExecutor = mailboxExecutor;
+		}
+
+		@Override
+		public void processElement(StreamRecord<T> element) throws Exception {
+			long timeout = (chainIndex + 1) * 1000L + count * 100L;
+			ProcessingTimeService processingTimeService = getProcessingTimeService();
+			processingTimeService
+				.registerTimer(
+					processingTimeService.getCurrentProcessingTime() + timeout,
+					timestamp -> {
+						events.add("Timer:" + chainIndex + ":" + count);
+						--count;
+					});
+
+			++count;
+			output.collect(element);
```

Comment:
The test is relying on collect being blocking until the downstream operators (or else yield might be called in the first operator first). Tbh I'm not sure if this a given or an implementation detail. So please double-check if that assumption always holds.

In any case I'd add a comment here. The test seemed rather random in the beginning, but it got much clearer after I realized this blocking behavior.

Reply:
Yes, I guess that's a fair assumption and if it is indeed changed somewhere in the future, then we will just adjust the test.